### PR TITLE
WIP - import

### PR DESCRIPTION
**Pull request by kunal-mohta**
This under development right now. 
There are several things left to be done, like error handling, etc.
I just want to know whether I am going in the right direction.

I want you to you to call the import statement at the top of a sketch with a valid sketch key and then check the console for an object.

### CODE REVIEW

**Comment by stebanos** [comment]
Thanks Kunal. I'm going to need some time to go through all this.

**Comment by kunal-mohta** [comment]
Sure!

File: js/core.js [hunk 1]
```diff
@@ -1046,9 +1047,39 @@ function parsePreamble(preamble, key, value, lineno) {
...
+            }
+            // else if(trimmedLine.length !== 0){
+            //     throw new Error(`Line ${ i+1 }: expecting '#import "sketch key" as reference-name'`);
+            // }
+            else return false;
+        }
+    }
+}
+
+function parsePhraseBook(s,importcode) {
```
Comment (stebanos) [inline js/core.js:1077]: Ok, if I understand correctly, you are pre-parsing the code looking for an import statement, and then load the sketch and the feed this to the parser. I don't think this is the way to go. There could be more than one 'import' statement. And an imported sketch could itself import another sketch.

File: js/core.js [hunk 2]
```diff
@@ -113,6 +113,7 @@ const ALPHANUM = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789
 const PREAMBLE_RE = /^\s*(\w+)\s*:\s*(.+)*$/;
 const POS_INTEGER_RE = /^\d+$/;
 const DURATION_RE = /^(\d+(\.\d+)?)\s*(s|ms)?$/;
+const IMPORT_RE = /^(\w+)\s*\"(.+)*\"\s*(as)\s*(\w+)$/;
```
Comment (stebanos) [inline js/core.js:116]: ` /^\s*import\s*\"(.+)\"\s*as\s*(\w+)\s*$/ ` would leave less error.

File: js/editor.js [hunk 3]
```diff
@@ -31,8 +31,18 @@ gMarkedRenderer.code = function(code, lang) {
         let result;
         try {
             console.log('rendering', code);
-            const phraseBook = parsePhraseBook(code);
-            result = generateString(phraseBook);
+            
+
+            if(checkImport(code)){
```
Comment (stebanos) [inline js/editor.js:36]: You are calling checkImport twice here.

File: js/core.js [hunk 4]
```diff
@@ -1067,11 +1098,29 @@ function parsePhraseBook(s) {
...
+                }
+                console.log(imported);
+            }
+            else if(trimmedLine.length !== 0){
+                //throw new Error(`Line ${ i+1 }: expecting '#import "sketch key" as reference-name'`);
+            } 
+            continue;
+        }
+        else if (trimmedLine[0] === '#') {
             // Ignore comments
```
Comment (stebanos) [inline js/core.js:1119]: I think checking for import statements should be part of this block.

**Review by stebanos — COMMENTED** [review]
Since it is in parsePhraseBook that imports are encountered, it should probably become a function that returns a Promise and and any consecutive action (like updating the GUI) can only happen only when everything is loaded and ready.

**Comment by stebanos** [comment]
Thank you again for your work Kunal! I wrote a few comments. Oh and, please take your time, we are not in a rush to having these issues solved =)

**Comment by kunal-mohta** [comment]
Thank you for pointing these out. 
I knew there were many issues in the code I wrote, but I needed to be sure that I am going in the right way.
I will review your comments and make changes accordingly.

**Comment by stebanos** [comment]
Hey Kunal, I'm closing this PR, since we already have a base to work from, ok?

**Comment by kunal-mohta** [comment]
Yes, sure!

**Pull request closed without merging**